tests/app/main/helpers/test_frameworks.py

```diff
@@ -638,3 +641,131 @@ def test_calls_logger_if_provided(self, abort):
             )
         ]
+
+
+class TestEnsureApplicationCompanyDetailsHaveBeenConfirmed(BaseApplicationTest):
+    def setup_method(self, method):
+        super().setup_method(method)
+
+        from app import data_api_client
+        self.data_api_client_mock = mock.Mock(spec_set=data_api_client)
+        self.decorator = EnsureApplicationCompanyDetailsHaveBeenConfirmed(self.data_api_client_mock)
+        # self.login()
+
+    def test_validator_raises_500_if_framework_slug_missing(self):
+        @EnsureApplicationCompanyDetailsHaveBeenConfirmed(self.data_api_client_mock)
+        def some_func():
+            pass
+
+        with mock.patch('app.main.helpers.frameworks.current_app') as current_app_patch:
+            with pytest.raises(HTTPException) as e:
+                some_func()
+
+        assert e.value.code == 500
+        assert current_app_patch.logger.error.call_args_list == [
+            mock.call("Required parameter `framework_slug` is undefined for the calling view.")
+        ]
+
+    def test_validator_raises_400_if_application_company_details_not_confirmed(self):
+        @EnsureApplicationCompanyDetailsHaveBeenConfirmed(self.data_api_client_mock)
+        def some_func(framework_slug):
+            pass
+
+        self.data_api_client_mock.get_supplier_framework_info.return_value = api_stubs.supplier_framework(
+            framework_slug='g-cloud-10',
+            application_company_details_confirmed=False
+        )
+
+        with mock.patch('app.main.helpers.frameworks.current_user') as current_user_patch:
+            current_user_patch.is_authenticated.return_value = True
+            current_user_patch.supplier_id.return_value = 1
+
+            with pytest.raises(HTTPException) as e:
+                some_func(framework_slug='g-cloud-10')
+
+        assert e.value.code == 400
+
+    def test_validator_returns_true_if_application_company_details_are_confirmed(self):
+        decorator = EnsureApplicationCompanyDetailsHaveBeenConfirmed(self.data_api_client_mock)
+
+        self.data_api_client_mock.get_supplier_framework_info.return_value = api_stubs.supplier_framework(
+            framework_slug='g-cloud-10',
+            application_company_details_confirmed=True
+        )
+
+        with mock.patch('app.main.helpers.frameworks.current_user') as current_user_patch:
+            current_user_patch.is_authenticated.return_value = True
+            current_user_patch.supplier_id.return_value = 1
+
+            assert decorator.validator(framework_slug='g-cloud-10') is True
+
+    def test_validator_can_be_reassigned(self):
+        def raise_successful_override():
+            return 'overriden'
+
+        decorator = EnsureApplicationCompanyDetailsHaveBeenConfirmed(self.data_api_client_mock)
+
+        with mock.patch.object(decorator, 'validator', raise_successful_override):
+            assert decorator.validator() == 'overriden'
+
+    def test_only_known_decorated_views_are_protected(self):
```

Comment:
I like this. My only concern was the overhead of having to call every view. But with (currently) 85 views it doesn't seem to be bad at all.

Reply:
👍

Reply:
I guess a slight thought is whether this test should be inverted to list the views which _shouldn't_ be protected explicitly, as that works on the principle that things are protected by default, which is generally more secure. There would be slightly more overhead in that every view which is added to this app (that isn't part of the application process, which is probably 99% of new views) would need to update this test.

Thoughts..?

Reply:
Interesting. I guess it would guard against an application view being added and not protected. It'd mean maintaining a longer list here in the test (65ish views).

Application stuff is all behind (I think) `/suppliers/frameworks/<framework_slug>`. Could you add some logic to the test that checks that any views behind that URL structure are protected? That way you could go some way to mitigate an application view being added and accidentally not protected, without having to list all the views here?